verilog/analysis/checkers/numeric_format_string_style_rule_test.cc

```diff
@@ -0,0 +1,75 @@
+
+#include "gtest/gtest.h"
+#include "common/analysis/linter_test_utils.h"
+#include "common/analysis/token_stream_linter_test_utils.h"
+#include "common/text/symbol.h"
+#include "verilog/analysis/verilog_analyzer.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+namespace analysis {
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+// Tests that the numeric formatting string is style-compilant
+TEST(NumericFormatStringStyleRuleTest, BasicTests) {
+  constexpr int kToken = TK_StringLiteral;
+  const std::initializer_list<LintTestCase> kTestCases = {
+      {""},
+
+      // hexadecimal value
+      {"module test; initial $display(", {kToken, "\"%0x\""}, ", some_hex_value); endmodule"},
+      {"module test; initial $display(\"0x%0x\", some_hex_value); endmodule"},
+      {"module test; initial $display(\"'h%0h\", some_hex_value); endmodule"},
+      {"module test; initial $display(\"Value: 0x%0x\", some_hex_value); endmodule"},
+      {"module test; initial $display(\"Value: 'x%0x\", some_hex_value); endmodule"},
+      {"module test; initial $display(", {kToken, "\"Value: 0X%0x\""}, ", some_hex_value); endmodule"},
+      {"module test; initial $display(", {kToken, "\"Value: 'X%0x\""}, ", some_hex_value); endmodule"},
+      {"module test; initial $display(", {kToken, "\"Value: %0x\""}, ", some_hex_value); endmodule"},
+
+      // binary value
+      {"module test; initial $display(", {kToken, "\"%0b\""}, ", some_bin_value); endmodule"},
+      {"module test; initial $display(\"0b%0b\", some_binary_value); endmodule"},
+      {"module test; initial $display(\"'b%0b\", some_binary_value); endmodule"}, // also acceptable
+      {"module test; initial $display(\"Value: 0b%0b\", some_binary_value); endmodule"},
+      {"module test; initial $display(\"Value: 'b%0b\", some_binary_value); endmodule"},
+      {"module test; initial $display(", {kToken, "\"Value: 0B%0b\""}, ", some_binary_value); endmodule"},
+      {"module test; initial $display(", {kToken, "\"Value: 'B%0b\""}, ", some_binary_value); endmodule"},
+      {"module test; initial $display(", {kToken, "\"Value: %0b\""}, ", some_binary_value); endmodule"},
+
+      // decimal value
+      {"module test; initial $display(\"%0d\", some_decimal_value); endmodule"}, // no prefix wanted
+      {"module test; initial $display(", {kToken, "\"0d%0d\""}, ", some_dec_value); endmodule"},
+      {"module test; initial $display(", {kToken, "\"'d%0d\""}, ", some_dec_value); endmodule"},
+      {"module test; initial $display(\"Value: %0d\", some_decimal_value); endmodule"}, // no prefix wanted
+      {"module test; initial $display(", {kToken, "\"Value: 0d%0d\""}, ", some_dec_value); endmodule"},
+      {"module test; initial $display(", {kToken, "\"Value: 'd%0d\""}, ", some_dec_value); endmodule"},
+
+      // mixed
+      {"module test; initial $display(", {kToken, "\"0x%0x, %x\""}, ", some_hex_value); endmodule"},
+  };
```

Comment:
Suggested test cases:

* single string literal with multiple violations inside
* string literal outside of a $display call (one case enough), like `parameter string foo = "...";`
* string literal inside $sformat call (one case enough)

Bonus:

* string literals inside macro definition bodies, and macro call arguments

To do this you'll want to to recursively lex `token.text()` like done here:
https://cs.opensource.google/verible/verible/+/master:verilog/analysis/verilog_equivalence.cc;l=92

Reply:
New test cases look good.  Do you plan work the macro definition bodies and call arguments, or shall we take this as-is and save that for later?